stan/math/prim/fun/transpose.hpp

```diff
@@ -1,13 +1,21 @@
 #ifndef STAN_MATH_PRIM_FUN_TRANSPOSE_HPP
 #define STAN_MATH_PRIM_FUN_TRANSPOSE_HPP
 
+#include <stan/math/prim/meta.hpp>
 #include <stan/math/prim/fun/Eigen.hpp>
 
 namespace stan {
 namespace math {
 
-template <typename T, int R, int C>
-Eigen::Matrix<T, C, R> inline transpose(const Eigen::Matrix<T, R, C>& m) {
+/**
+ * Transposes a matrix.
+ * @tparam T type of the matrix or expression
+ * @param m matrix or expression
+ * @return transposed matrix
+ */
+template <typename T, typename = require_eigen_t<T>>
+Eigen::Matrix<value_type_t<T>, T::ColsAtCompileTime,
+              T::RowsAtCompileTime> inline transpose(const T& m) {
```

Comment:
We can't just use .eval() or .matrix() here?

Reply:
You mean to use auto? Unfortunately not yet. `m.transpose().eval()` returns a row major matrix. This would not work with other functions untill they are generalized.